app/models/user_notification.rb

```diff
@@ -0,0 +1,42 @@
+# The user notifications for activities. This is meant to be called by the activities framework
+# alone.
+#
+# @api private
+class UserNotification < ActiveRecord::Base
+  acts_as_readable on: :created_at
+  after_create :send_email
+
+  belongs_to :activity, inverse_of: :user_notifications
+  belongs_to :user, inverse_of: :notifications
+
+  enum notification_type: { popup: 0, email: 1 }
+  SUPPORTED_TYPES = [:popup, :email]
+
+  # Create notification based on a given activity.
+  #
+  # @param [Activity] activity the activity which stores main information of this notification.
+  # @param [User] user the user who will receive notifications.
+  # @param [Array<Symbol>] types the formats of the notifications.
+  #   Can be :popup, :email or both.
+  def self.create(activity, user, types)
+    types.each do |type|
+      fail 'Invalid type is used to create a notification!' unless SUPPORTED_TYPES.include?(type)
+      super({ activity: activity, user: user, notification_type: type })
+    end
```

Comment:
Same here, you don't need override this function.  And you are using a loop to create multiple objects in one `create`, which I think is wrong. 

You let caller(i.e. `notify` here ) to iterate through each type.